### ISSUE TITLE
[=BUG=] Legacy videos handeling problem

### ISSUE THREAD

**Issue by Hydro001** [play-dl/play-dl#258]
#### Describe the bug

 - Node.js version: 17.5.0

Bot crashing when trying to stream a "legacy" video.

#### Actual behavior

Successfully shows the stream type, sends an error in the console and crashes the bot
#### Expected behavior

 
Work normally or atleast being able to do a try-catch to prevent the bot from crashing
#### Code to reproduce

```js
const playStream = await play.stream('https://www.youtube.com/watch?v=aA6zYJNPWXI');
console.log(playStream.type);
```

#### ScreenShots

```
arbitrary
C:\Users\User\Documents\Discord Bots\Bot\node_modules\play-dl\dist\index.js:7
${c[2].playerResponse.playabilityStatus.errorScreen.playerErrorMessageRenderer?.reason.simpleText??c[2].playerResponse.playabilityStatus.errorScreen.playerKavRenderer?.reason.simpleText}`);let p=c[2].playerResponse.streamingData;return r?{streamingData:p,relatedVideos:c[3].response.contents.twoColumnWatchNextResults.secondaryResults}:{streamingData:p}}a(tt,"acceptViewerDiscretion");function ri(t,e,i){let r=t.contents.twoColumnWatchNextResults.playlist.playlist,s=ni(r.contents),n=e.split('INNERTUBE_API_KEY":"')[1]?.split('"')[0]??e.split('innertubeApiKey":"')[1]?.split('"')[0]??Ee,o=r.totalVideos,l=r.shortBylineText?.runs?.[0],u=r.badges?.[0]?.metadataBadgeRenderer?.style.toLowerCase();return new D({continuation:{api:n,token:ie(r.contents),clientVersion:e.split('"INNERTUBE_CONTEXT_CLIENT_VERSION":"')[1]?.split('"')[0]??e.split('"innertube_context_client_version":"')[1]?.split('"')[0]??"<some version>"},id:r.playlistId||"",title:r.title||"",videoCount:parseInt(o)||0,videos:s,url:i,channel:{id:l?.navigationEndpoint?.browseEndpoint?.browseId||null,name:l?.text||null,url:`https://www.youtube.com${l?.navigationEndpoint?.browseEndpoint?.canonicalBaseUrl||l?.navigationEndpoint?.commandMetadata?.webCommandMetadata?.url}`,verified:Boolean(u?.includes("verified")),artist:Boolean(u?.includes("artist"))}})}a(ri,"getWatchPlaylist");function si(t,e){let i=t.contents.twoColumnBrowseResultsRenderer.tabs[0].tabRenderer.content.sectionListRenderer.contents[0].itemSectionRenderer.contents[0].playlistVideoListRenderer.contents,r=t.sidebar.playlistSidebarRenderer.items,s=e.split('INNERTUBE_API_KEY":"')[1]?.split('"')[0]??e.split('innertubeApiKey":"')[1]?.split('"')[0]??Ee,n=Te(i,100),o=r[0].playlistSidebarPrimaryInfoRenderer;if(!o.title.runs||!o.title.runs.length)throw new Error("Failed to Parse Playlist info.");let l=r[1]?.playlistSidebarSecondaryInfoRenderer.videoOwner,u=o.stats.length===3?o.stats[1].simpleText.replace(/\D/g,""):0,c=o.stats.find(y=>"runs"in y&&y.runs.find(_=>_.text.toLowerCase().includes("last update")))?.runs.pop()?.text??null,p=o.stats[0].runs[0].text.replace(/\D/g,"")||0;return new D({continuation:{api:s,token:ie(i),clientVersion:e.split('"INNERTUBE_CONTEXT_CLIENT_VERSION":"')[1]?.split('"')[0]??e.split('"innertube_context_client_version":"')[1]?.split('"')[0]??"<some version>"},id:o.title.runs[0].navigationEndpoint.watchEndpoint.playlistId,title:o.title.runs[0].text,videoCount:parseInt(p)||0,lastUpdate:c,views:parseInt(u)||0,videos:n,url:`https://www.youtube.com/playlist?list=${o.title.runs[0].navigationEndpoint.watchEndpoint.playlistId}`,link:`https://www.youtube.com${o.title.runs[0].navigationEndpoint.commandMetadata.webCommandMetadata.url}`,channel:l?{name:l.videoOwnerRenderer.title.runs[0].text,id:l.videoOwnerRenderer.title.runs[0].navigationEndpoint.browseEndpoint.browseId,url:`https://www.youtube.com${l.videoOwnerRenderer.navigationEndpoint.commandMetadata.webCommandMetadata.url||l.videoOwnerRenderer.navigationEndpoint.browseEndpoint.canonicalBaseUrl}`,icons:l.videoOwnerRenderer.thumbnail.thumbnails??[]}:{},thumbnail:o.thumbnailRenderer.playlistVideoThumbnailRenderer?.thumbnail.thumbnails.length?o.thumbnailRenderer.playlistVideoThumbnailRenderer.thumbnail.thumbnails[o.thumbnailRenderer.playlistVideoThumbnailRenderer.thumbnail.thumbnails.length-1]:null})}a(si,"getNormalPlaylist");function ni(t,e=1/0){let i=[];for(let r=0;r<t.length&&e!==i.length;r++){let s=t[r].playlistPanelVideoRenderer;if(!s||!s.shortBylineText)continue;let n=s.shortBylineText.runs[0];i.push(new T({id:s.videoId,duration:oi(s.lengthText?.simpleText)||0,duration_raw:s.lengthText?.simpleText??"0:00",thumbnails:s.thumbnail.thumbnails,title:s.title.simpleText,upcoming:s.thumbnailOverlays[0].thumbnailOverlayTimeStatusRenderer.style==="UPCOMING"||void 0,channel:{id:n.navigationEndpoint.browseEndpoint.browseId||void 0,name:n.text||void 0,url:`https://www.youtube.com${n.navigationEndpoint.browseEndpoint.canonicalBaseUrl||n.navigationEndpoint.commandMetadata.webCommandMetadata.url}`,icon:void 0}}))}return i}a(ni,"getWatchPlaylistVideos");function oi(t){if(!t)return 0;let e=t.split(":");switch(e.length){case 2:return parseInt(e[0])*60+parseInt(e[1]);case 3:return parseInt(e[0])*60*60+parseInt(e[1])*60+parseInt(e[2]);default:return 0}}a(oi,"parseDuration");var it=require("url"),le=class{constructor(e,i,r,s){this.stream=new De.Readable({highWaterMark:5*1e3*1e3,read(){}}),this.type="arbitrary",this.sequence=0,this.dash_url=e,this.base_url="",this.interval=i,this.video_url=r,this.precache=s||3,this.dash_timer=new O(()=>{this.dash_updater(),this.dash_timer.reuse()},1800),this.stream.on("close",()=>{this.cleanup()}),this.initialize_dash()}cleanup(){this.normal_timer?.destroy(),this.dash_timer.destroy(),this.request?.destroy(),this.video_url="",this.request=void 0,this.dash_url="",this.base_url="",this.interval=0}async dash_updater(){let e=await B(this.video_url);return e.LiveStreamData.dashManifestUrl&&(this.dash_url=e.LiveStreamData.dashManifestUrl),this.initialize_dash()}async initialize_dash(){let i=(await h(this.dash_url)).split('<AdaptationSet id="0"')[1].split("</AdaptationSet>")[0].split("</Representation>");if(i[i.length-1]===""&&i.pop(),this.base_url=i[i.length-1].split("<BaseURL>")[1].split("</BaseURL>")[0],await w(`https://${new it.URL(this.base_url).host}/generate_204`),this.sequence===0){let r=i[i.length-1].split("<SegmentList>")[1].split("</SegmentList>")[0].replaceAll('<SegmentURL media="',"").split('"/>');r[r.length-1]===""&&r.pop(),r.length>this.precache&&r.splice(0,r.length-this.precache),this.sequence=Number(r[0].split("sq/")[1].split("/")[0]),this.first_data(r.length)}}async first_data(e){for(let i=1;i<=e;i++)await new Promise(async r=>{let s=await w(this.base_url+"sq/"+this.sequence).catch(n=>n);if(s instanceof Error){this.stream.emit("error",s);return}this.request=s,s.on("data",n=>{this.stream.push(n)}),s.on("end",()=>{this.sequence++,r("")}),s.once("error",n=>{this.stream.emit("error",n)})});this.normal_timer=new O(()=>{this.loop(),this.normal_timer?.reuse()},this.interval)}loop(){return new Promise(async e=>{let i=await w(this.base_url+"sq/"+this.sequence).catch(r=>r);if(i instanceof Error){this.stream.emit("error",i);return}this.request=i,i.on("data",r=>{this.stream.push(r)}),i.on("end",()=>{this.sequence++,e("")}),i.once("error",r=>{this.stream.emit("error",r)})})}pause(){}resume(){}};a(le,"LiveStream");var ue=class{constructor(e,i,r,s,n,o){this.stream=new De.Readable({highWaterMark:5*1e3*1e3,read(){}}),this.url=e,this.quality=o.quality,this.type=i,this.bytes_count=0,this.video_url=n,this.per_sec_bytes=Math.ceil(s/r),this.content_length=s,this.request=null,this.timer=new O(()=>{this.timer.reuse(),this.loop()},265),this.stream.on("close",()=>{this.timer.destroy(),this.cleanup()}),this.loop()}async retry(){let e=await B(this.video_url),i=J(e.format);this.url=i[this.quality].url}cleanup(){this.request?.destroy(),this.request=null,this.url=""}async loop(){if(this.stream.destroyed){this.timer.destroy(),this.cleanup();return}let e=this.bytes_count+this.per_sec_bytes*300,i=await w(this.url,{headers:{range:`bytes=${this.bytes_count}-${e>=this.content_length?"":e}`}}).catch(r=>r);if(i instanceof Error){this.stream.emit("error",i),this.bytes_count=0,this.per_sec_bytes=0,this.cleanup();return}if(Number(i.statusCode)>=400){this.cleanup(),await this.retry(),this.timer.reuse(),this.loop();return}this.request=i,i.on("data",r=>{this.stream.push(r)}),i.once("error",async()=>{this.cleanup(),await this.retry(),this.timer.reuse(),this.loop()}),i.on("data",r=>{this.bytes_count+=r.length}),i.on("end",()=>{e>=this.content_length&&(this.timer.destroy(),this.stream.push(null),this.cleanup())})}pause(){this.timer.pause()}resume(){this.timer.resume()}};a(ue,"Stream");var O=class{constructor(e,i){this.callback=e,this.time_total=i,this.time_left=i,this.paused=!1,this.destroyed=!1,this.time_start=process.hrtime()[0],this.timer=setTimeout(this.callback,this.time_total*1e3)}pause(){return!this.paused&&!this.destroyed?(this.paused=!0,clearTimeout(this.timer),this.time_left=this.time_left-(process.hrtime()[0]-this.time_start),!0):!1}resume(){return this.paused&&!this.destroyed?(this.paused=!1,this.time_start=process.hrtime()[0],this.timer=setTimeout(this.callback,this.time_left*1e3),!0):!1}reuse(){return this.destroyed?!1:(clearTimeout(this.timer),this.time_left=this.time_total,this.paused=!1,this.time_start=process.hrtime()[0],this.timer=setTimeout(this.callback,this.time_total*1e3),!0)}destroy(){clearTimeout(this.timer),this.destroyed=!0,this.callback=()=>{},this.time_total=0,this.time_left=0,this.paused=!1,this.time_start=0}};a(O,"Timer");var Z=require("play-audio"),rt=require("stream");var ai=Object.keys(Z.WebmElements),ce=class extends rt.Duplex{constructor(e,i){super(i);this.state="READING_HEAD",this.cursor=0,this.header=new Z.WebmHeader,this.headfound=!1,this.headerparsed=!1,this.seekfound=!1,this.data_length=0,this.data_size=0,this.offset=0,this.sec=e,this.time=Math.floor(e/10)*10}get vint_length(){let e=0;for(;e<8&&!(1<<7-e&this.chunk[this.cursor]);e++);return++e}vint_value(){if(!this.chunk)return!1;let e=this.vint_length;if(this.chunk.length<this.cursor+e)return!1;let i=this.chunk[this.cursor]&(1<<8-e)-1;for(let r=this.cursor+1;r<this.cursor+e;r++)i=(i<<8)+this.chunk[r];return this.data_size=e,this.data_length=i,!0}cleanup(){this.cursor=0,this.chunk=void 0,this.remaining=void 0}_read(){}seek(e){let i=0,r=0,s=(this.sec-this.time)*1e3||0;if(s=Math.round(s/20)*20,!this.header.segment.cues)return new Error("Failed to Parse Cues");for(let n=0;n<this.header.segment.cues.length;n++){let o=this.header.segment.cues[n];if(Math.floor(o.time/1e3)===this.time){r=o.position,i=(this.header.segment.cues[n+1]?.position||e)-r-1;break}else continue}return i===0?r:this.offset+Math.round(r+s/20*(i/500))}_write(e,i,r){this.remaining?(this.chunk=Buffer.concat([this.remaining,e]),this.remaining=void 0):this.chunk=e;let s;this.state==="READING_HEAD"?s=this.readHead():this.seekfound?s=this.readTag():s=this.getClosestBlock(),s?r(s):r()}readHead(){if(!this.chunk)return new Error("Chunk is missing");for(;this.chunk.length>this.cursor;){let e=this.cursor,i=this.vint_length;if(this.chunk.length<this.cursor+i)break;let r=this.parseEbmlID(this.chunk.slice(this.cursor,this.cursor+i).toString("hex"));if(this.cursor+=i,!this.vint_value()){this.cursor=e;break}if(!r){this.cursor+=this.data_size+this.data_length;continue}if(!this.headfound)if(r.name==="ebml")this.headfound=!0;else return new Error("Failed to find EBML ID at start of stream.");let s=this.chunk.slice(this.cursor+this.data_size,this.cursor+this.data_size+this.data_length),n=this.header.parse(r,s);if(n instanceof Error)return n;if(r.name==="seekHead"&&(this.offset=e),r.name==="cueClusterPosition"&&this.header.segment.cues.length>2&&this.time===this.header.segment.cues.at(-2).time/1e3&&this.emit("headComplete"),r.type===0){this.cursor+=this.data_size;continue}if(this.chunk.length<this.cursor+this.data_size+this.data_length){this.cursor=e;break}else this.cursor+=this.data_size+this.data_length}this.remaining=this.chunk.slice(this.cursor),this.cursor=0}readTag(){if(!this.chunk)return new Error("Chunk is missing");for(;this.chunk.length>this.cursor;){let e=this.cursor,i=this.vint_length;if(this.chunk.length<this.cursor+i)break;let r=this.parseEbmlID(this.chunk.slice(this.cursor,this.cursor+i).toString("hex"));if(this.cursor+=i,!this.vint_value()){this.cursor=e;break}if(!r){this.cursor+=this.data_size+this.data_length;continue}let s=this.chunk.slice(this.cursor+this.data_size,this.cursor+this.data_size+this.data_length),n=this.header.parse(r,s);if(n instanceof Error)return n;if(r.type===0){this.cursor+=this.data_size;continue}if(this.chunk.length<this.cursor+this.data_size+this.data_length){this.cursor=e;break}else this.cursor+=this.data_size+this.data_length;if(r.name==="simpleBlock"){let o=this.header.segment.tracks[this.header.audioTrack];if(!o||o.trackType!==2)return new Error("No audio Track in this webm file.");(s[0]&15)===o.trackNumber&&this.push(s.slice(4))}}this.remaining=this.chunk.slice(this.cursor),this.cursor=0}getClosestBlock(){if(this.sec===0)return this.seekfound=!0,this.readTag();if(!this.chunk)return new Error("Chunk is missing");this.cursor=0;let e=!1;for(;!e&&this.cursor<this.chunk.length;){if(this.cursor=this.chunk.indexOf("a3",this.cursor,"hex"),this.cursor===-1)return new Error("Failed to find nearest Block.");if(this.cursor++,!this.vint_value())return new Error("Failed to find correct simpleBlock in first chunk");if(this.cursor+this.data_length+this.data_length>this.chunk.length)continue;let i=this.chunk.slice(this.cursor+this.data_size,this.cursor+this.data_size+this.data_length),r=this.header.segment.tracks[this.header.audioTrack];if(!r||r.trackType!==2)return new Error("No audio Track in this webm file.");if((i[0]&15)===r.trackNumber)this.cursor+=this.data_size+this.data_length,this.push(i.slice(4)),e=!0;else continue}return e?(this.seekfound=!0,this.readTag()):new Error("Failed to find nearest correct simple Block.")}parseEbmlID(e){return ai.includes(e)?Z.WebmElements[e]:!1}_destroy(e,i){this.cleanup(),i(e)}_final(e){this.cleanup(),e()}};a(ce,"WebmSeeker");var he=class{constructor(e,i,r,s,n,o){this.stream=new ce(o.seek,{highWaterMark:5*1e3*1e3,readableObjectMode:!0}),this.url=e,this.quality=o.quality,this.type="opus",this.bytes_count=0,this.video_url=n,this.per_sec_bytes=Math.ceil(s/i),this.header_length=r,this.content_length=s,this.request=null,this.timer=new O(()=>{this.timer.reuse(),this.loop()},265),this.stream.on("close",()=>{this.timer.destroy(),this.cleanup()}),this.seek()}async seek(){let e=await new Promise(async(r,s)=>{if(this.stream.headerparsed)r("");else{let n=await w(this.url,{headers:{range:`bytes=0-${this.header_length}`}}).catch(o=>o);if(n instanceof Error){s(n);return}if(Number(n.statusCode)>=400){s(400);return}this.request=n,n.pipe(this.stream,{end:!1}),n.once("end",()=>{this.stream.state="READING_DATA",r("")}),this.stream.once("headComplete",()=>{n.unpipe(this.stream),n.destroy(),this.stream.state="READING_DATA",r("")})}}).catch(r=>r);if(e instanceof Error){this.stream.emit("error",e),this.bytes_count=0,this.per_sec_bytes=0,this.cleanup();return}else if(e===400)return await this.retry(),this.timer.reuse(),this.seek();let i=this.stream.seek(this.content_length);if(i instanceof Error){this.stream.emit("error",i),this.bytes_count=0,this.per_sec_bytes=0,this.cleanup();return}this.stream.seekfound=!1,this.bytes_count=i,this.timer.reuse(),this.loop()}async retry(){let e=await B(this.video_url),i=J(e.format);this.url=i[this.quality].url}cleanup(){this.request?.destroy(),this.request=null,this.url=""}async loop(){if(this.stream.destroyed){this.timer.destroy(),this.cleanup();return}let e=this.bytes_count+this.per_sec_bytes*300,i=await w(this.url,{headers:{range:`bytes=${this.bytes_count}-${e>=this.content_length?"":e}`}}).catch(r=>r);if(i instanceof Error){this.stream.emit("error",i),this.bytes_count=0,this.per_sec_bytes=0,this.cleanup();return}if(Number(i.statusCode)>=400){this.cleanup(),await this.retry(),this.timer.reuse(),this.loop();return}this.request=i,i.pipe(this.stream,{end:!1}),i.once("error",async()=>{this.cleanup(),await this.retry(),this.timer.reuse(),this.loop()}),i.on("data",r=>{this.bytes_count+=r.length}),i.on("end",()=>{e>=this.content_length&&(this.timer.destroy(),this.stream.end(),this.cleanup())})}pause(){this.timer.pause()}resume(){this.timer.resume()}};a(he,"SeekStream");var st=require("url");function J(t){let e=[];return t.forEach(i=>{let r=i.mimeType;r.startsWith("audio")&&(i.codec=r.split('codecs="')[1].split('"')[0],i.container=r.split("audio/")[1].split(";")[0],e.push(i))}),e}a(J,"parseAudioFormats");async function de(t,e={}){let i=await B(t,{htmldata:e.htmldata,language:e.language});return await pe(i,e)}a(de,"stream");async function pe(t,e={}){if(t.format.length===0)throw new Error("Upcoming and premiere videos that are not currently live cannot be streamed.");let i=[];if(t.LiveStreamData.isLive===!0&&t.LiveStreamData.dashManifestUrl!==null&&t.video_details.durationInSec===0)return new le(t.LiveStreamData.dashManifestUrl,t.format[t.format.length-1].targetDurationSec,t.video_details.url,e.precache);let r=J(t.format);typeof e.quality!="number"?e.quality=r.length-1:e.quality<=0?e.quality=0:e.quality>=r.length&&(e.quality=r.length-1),r.length!==0?i.push(r[e.quality]):i.push(t.format[t.format.length-1]);let s=i[0].codec==="opus"&&i[0].container==="webm"?"webm/opus":"arbitrary";if(await w(`https://${new st.URL(i[0].url).host}/generate_204`),s==="webm/opus")if(e.discordPlayerCompatibility){if(e.seek)throw new Error("Can not seek with discordPlayerCompatibility set to true.")}else{if(e.seek??=0,e.seek>=t.video_details.durationInSec||e.seek<0)throw new Error(`Seeking beyond limit. [ 0 - ${t.video_details.durationInSec-1}]`);return new he(i[0].url,t.video_details.durationInSec,i[0].indexRange.end,Number(i[0].contentLength),t.video_details.url,e)}return new ue(i[0].url,s,t.video_details.durationInSec,Number(i[0].contentLength),t.video_details.url,e)}a(pe,"stream_from_info");var li=["-oaymwEpCOADEI4CSFryq4qpAxsIARUAAAAAGAElAADIQj0AgKJDeAHtAZmZGUI=","-oaymwEiCOADEI4CSFXyq4qpAxQIARUAAIhCGAFwAcABBu0BmZkZQg==","-oaymwEiCOgCEMoBSFXyq4qpAxQIARUAAIhCGAFwAcABBu0BZmbmQQ==","-oaymwEiCNAFEJQDSFXyq4qpAxQIARUAAIhCGAFwAcABBu0BZmZmQg==","-oaymwEdCNAFEJQDSFryq4qpAw8IARUAAIhCGAHtAWZmZkI=","-oaymwEdCNACELwBSFryq4qpAw8IARUAAIhCGAHtAT0K10E="];function ot(t,e){if(!t)throw new Error("Can't parse Search result without data");e?e.type||(e.type="video"):e={type:"video",limit:0};let i=typeof e.limit=="number"&&e.limit>0;e.unblurNSFWThumbnails??=!1;let r=t.split("var ytInitialData = ")?.[1]?.split(";<\/script>")[0].split(/;\s*(var|const|let)\s/)[0],s=JSON.parse(r),n=[],o=s.contents.twoColumnSearchResultsRenderer.primaryContents.sectionListRenderer.contents[0].itemSectionRenderer.contents;for(let l of o){if(i&&n.length===e.limit)break;if(!(!l.videoRenderer&&!l.channelRenderer&&!l.playlistRenderer))switch(e.type){case"video":{let u=hi(l);u&&(e.unblurNSFWThumbnails&&u.thumbnails.forEach(nt),n.push(u));break}case"channel":{let u=ci(l);u&&n.push(u);break}case"playlist":{let u=di(l);u&&(e.unblurNSFWThumbnails&&u.thumbnail&&nt(u.thumbnail),n.push(u));break}default:throw new Error(`Unknown search type: ${e.type}`)}}return n}a(ot,"ParseSearchResult");function ui(t){if(!t)return 0;let e=t.split(":"),i=0;switch(e.length){case 3:i=parseInt(e[0])*60*60+parseInt(e[1])*60+parseInt(e[2]);break;case 2:i=parseInt(e[0])*60+parseInt(e[1]);break;default:i=parseInt(e[0])}return i}a(ui,"parseDuration");function ci(t){if(!t||!t.channelRenderer)throw new Error("Failed to Parse YouTube Channel");let e=t.channelRenderer.ownerBadges?.[0]?.metadataBadgeRenderer?.style?.toLowerCase(),i=`https://www.youtube.com${t.channelRenderer.navigationEndpoint.browseEndpoint.canonicalBaseUrl||t.channelRenderer.navigationEndpoint.commandMetadata.webCommandMetadata.url}`,r=t.channelRenderer.thumbnail.thumbnails[t.channelRenderer.thumbnail.thumbnails.length-1];return new S({id:t.channelRenderer.channelId,name:t.channelRenderer.title.simpleText,icon:{url:r.url.replace("//","https://"),width:r.width,height:r.height},url:i,verified:Boolean(e?.includes("verified")),artist:Boolean(e?.includes("artist")),subscribers:t.channelRenderer.subscriberCountText?.simpleText??"0 subscribers"})}a(ci,"parseChannel");function hi(t){if(!t||!t.videoRenderer)throw new Error("Failed to 
Parse YouTube Video");let e=t.videoRenderer.ownerText.runs[0],i=t.videoRenderer.ownerBadges?.[0]?.metadataBadgeRenderer?.style?.toLowerCase(),r=t.videoRenderer.lengthText;return new T({id:t.videoRenderer.videoId,url:`https://www.youtube.com/watch?v=${t.videoRenderer.videoId}`,title:t.videoRenderer.title.runs[0].text,description:t.videoRenderer.detailedMetadataSnippets?.[0].snippetText.runs.length?t.videoRenderer.detailedMetadataSnippets[0].snippetText.runs.map(n=>n.text).join(""):"",duration:r?ui(r.simpleText):0,duration_raw:r?r.simpleText:null,thumbnails:t.videoRenderer.thumbnail.thumbnails,channel:{id:e.navigationEndpoint.browseEndpoint.browseId||null,name:e.text||null,url:`https://www.youtube.com${e.navigationEndpoint.browseEndpoint.canonicalBaseUrl||e.navigationEndpoint.commandMetadata.webCommandMetadata.url}`,icons:t.videoRenderer.channelThumbnailSupportedRenderers.channelThumbnailWithLinkRenderer.thumbnail.thumbnails,verified:Boolean(i?.includes("verified")),artist:Boolean(i?.includes("artist"))},uploadedAt:t.videoRenderer.publishedTimeText?.simpleText??null,upcoming:t.videoRenderer.upcomingEventData?.startTime?new Date(parseInt(t.videoRenderer.upcomingEventData.startTime)*1e3):void 0,views:t.videoRenderer.viewCountText?.simpleText?.replace(/\D/g,"")??0,live:!r})}a(hi,"parseVideo");function di(t){if(!t||!t.playlistRenderer)throw new Error("Failed to Parse YouTube Playlist");let e=t.playlistRenderer.thumbnails[0].thumbnails[t.playlistRenderer.thumbnails[0].thumbnails.length-1],i=t.playlistRenderer.shortBylineText.runs?.[0];return new D({id:t.playlistRenderer.playlistId,title:t.playlistRenderer.title.simpleText,thumbnail:{id:t.playlistRenderer.playlistId,url:e.url,height:e.height,width:e.width},channel:{id:i?.navigationEndpoint.browseEndpoint.browseId,name:i?.text,url:`https://www.youtube.com${i?.navigationEndpoint.commandMetadata.webCommandMetadata.url}`},videos:parseInt(t.playlistRenderer.videoCount.replace(/\D/g,""))},!0)}a(di,"parsePlaylist");function nt(t){if(li.find(e=>t.url.includes(e)))switch(t.url=t.url.split("?")[0],t.url.split("/").at(-1).split(".")[0]){case"hq2":case"hqdefault":t.width=480,t.height=360;break;case"hq720":t.width=1280,t.height=720;break;case"sddefault":t.width=640,t.height=480;break;case"mqdefault":t.width=320,t.height=180;break;case"default":t.width=120,t.height=90;break;default:t.width=t.height=NaN}}a(nt,"unblurThumbnail");async function at(t,e={}){let i="https://www.youtube.com/results?search_query="+t;if(e.type??="video",i.indexOf("&sp=")===-1)switch(i+="&sp=",e.type){case"channel":i+="EgIQAg%253D%253D";break;case"playlist":i+="EgIQAw%253D%253D";break;case"video":i+="EgIQAQ%253D%253D";break;default:throw new Error(`Unknown search type: ${e.type}`)}let r=await h(i,{headers:{"accept-language":e.language||"en-US;q=0.9"}});if(r.indexOf("Our systems have detected unusual traffic from your computer network.")!==-1)throw new Error("Captcha page: YouTube has detected that you are a bot!");return ot(r,e)}a(at,"yt_search");var E=class{constructor(e){this.name=e.name,this.id=e.id,this.type="track",this.url=e.external_urls.spotify,this.explicit=e.explicit,this.durationInMs=e.duration_ms,this.durationInSec=Math.round(this.durationInMs/1e3);let i=[];e.artists.forEach(r=>{i.push({name:r.name,id:r.id,url:r.external_urls.spotify})}),this.artists=i,e.album?.name?this.album={name:e.album.name,url:e.external_urls.spotify,id:e.album.id,release_date:e.album.release_date,release_date_precision:e.album.release_date_precision,total_tracks:e.album.total_tracks}:this.album=void 0,e.album?.images?.[0]?this.thumbnail=e.album.images[0]:this.thumbnail=void 0}toJSON(){return{name:this.name,id:this.id,url:this.url,explicit:this.explicit,durationInMs:this.durationInMs,durationInSec:this.durationInSec,artists:this.artists,album:this.album,thumbnail:this.thumbnail}}};a(E,"SpotifyTrack");var A=class{constructor(e,i,r){this.name=e.name,this.type="playlist",this.search=r,this.collaborative=e.collaborative,this.description=e.description,this.url=e.external_urls.spotify,this.id=e.id,this.thumbnail=e.images[0],this.owner={name:e.owner.display_name,url:e.owner.external_urls.spotify,id:e.owner.id},this.tracksCount=Number(e.tracks.total);let s=[];this.search||e.tracks.items.forEach(n=>{n.track&&s.push(new E(n.track))}),this.fetched_tracks=new Map,this.fetched_tracks.set("1",s),this.spotifyData=i}async fetch(){if(this.search)return this;let e;if(this.tracksCount>1e3?e=1e3:e=this.tracksCount,e<=100)return this;let i=[];for(let r=2;r<=Math.ceil(e/100);r++)i.push(new Promise(async(s,n)=>{let o=await h(`https://api.spotify.com/v1/playlists/${this.id}/tracks?offset=${(r-1)*100}&limit=100&market=${this.spotifyData.market}`,{headers:{Authorization:`${this.spotifyData.token_type} ${this.spotifyData.access_token}`}}).catch(c=>n(`Response Error :






TypeError: Cannot read properties of undefined (reading 'url')
    at ue.retry (C:\Users\User\Documents\Discord Bots\Bot\node_modules\play-dl\dist\index.js:7:6887)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async ue.loop (C:\Users\User\Documents\Discord Bots\Bot\node_modules\play-dl\dist\index.js:7:7350)
```
#### Checklist

- [✔️] I have read the documentation.
- [✔️] I have tried my code with the latest version of Node.js and Play-dl.

**Comment by absidue** [comment]
This should be fixed in 1.8.9.

**Comment by Hydro001** [comment]
It is, thank you ^^
